// @wdio/devtools-service/dist/index.js
import { C as e, S as t, _ as n, a as r, b as i, c as a, d as o, f as s, g as c, h as l, i as u, l as d, m as ee, n as f, o as p, p as te, r as m, s as ne, t as re, u as h, v as ie, x as ae, y as oe } from "./launcher-DpNc4Ovb.js";
import { TraceType as g } from "./types.js";
import _ from "node:fs/promises";
import v from "node:path";
import y from "@wdio/logger";
import { SevereServiceError as b } from "webdriverio";
import x from "node:url";
import { WebSocket as S } from "ws";
import { parse as C } from "stack-trace";
import { resolve as se } from "import-meta-resolve";
import ce from "@wdio/reporter";
import w from "fs";
import { createRequire as T } from "node:module";
import { parse as E } from "@babel/parser";
import { existsSync as D, readFileSync as O } from "node:fs";
//#region src/session.ts
var k = y("@wdio/devtools-service:SessionCapturer"), A = (e) => e.replace(f, ""), le = (e) => {
	let t = A(e).toLowerCase();
	for (let { level: e, pattern: n } of d) if (n.test(t)) return e;
	return r.some((e) => t.includes(e.toLowerCase())) ? "error" : "log";
}, j = (e, t, n) => ({
	timestamp: Date.now(),
	type: e,
	args: t,
	source: n
}), M = class {
	#e;
	#t = !1;
	#n;
	#r;
	#i = !1;
	#a = /* @__PURE__ */ new Map();
	constructor(e = {}) {
		this.commandsLog = [], this.sources = /* @__PURE__ */ new Map(), this.mutations = [], this.traceLogs = [], this.consoleLogs = [], this.networkRequests = [];
		let { port: t, hostname: n } = e;
		n && t && (this.#e = new S(`ws://${n}:${t}/worker`), this.#e.on("error", (e) => k.error(`Couldn't connect to devtools backend: ${e.message}`))), this.#n = {
			log: console.log,
			info: console.info,
			warn: console.warn,
			error: console.error
		}, this.#r = {
			stdoutWrite: process.stdout.write.bind(process.stdout),
			stderrWrite: process.stderr.write.bind(process.stderr)
		}, this.#o(), this.#s();
	}
	#o() {
		m.forEach((e) => {
			let t = this.#n[e];
			console[e] = (...n) => {
				let r = j(e, n.map((e) => typeof e == "object" && e ? (() => {
					try {
						return JSON.stringify(e);
					} catch {
						return String(e);
					}
				})() : String(e)), h.TEST);
				this.consoleLogs.push(r), this.sendUpstream("consoleLogs", [r]), this.#i = !0;
				let i = t.apply(console, n);
				return this.#i = !1, i;
			};
		});
	}
	#s() {
		let e = (e) => {
			let t = typeof e == "string" ? e : e.toString();
			t?.trim() && t.split("\n").filter((e) => e.trim()).forEach((e) => {
				let t = j(le(e), [A(e)], h.TERMINAL);
				this.consoleLogs.push(t), this.sendUpstream("consoleLogs", [t]);
			});
		}, t = (t, n) => {
			let r = this;
			t.write = function(i, ...a) {
				let o = n.call(t, i, ...a);
				return i && !r.#i && e(i), o;
			};
		};
		t(process.stdout, this.#r.stdoutWrite), t(process.stderr, this.#r.stderrWrite);
	}
	#c() {
		m.forEach((e) => {
			console[e] = this.#n[e];
		});
	}
	cleanup() {
		this.#c();
	}
	get isReportingUpstream() {
		return !!this.#e && this.#e?.readyState === S.OPEN;
	}
	async afterCommand(e, t, n, r, i, a) {
		let s = C(/* @__PURE__ */ Error("")).filter((e) => !!e.getFileName()).map((e) => [
			e.getFileName(),
			e.getLineNumber(),
			e.getColumnNumber()
		].join(":")).filter((e) => !e.includes("/node_modules/") && !e.includes("<anonymous>)") && !e.includes("node:internal") && !e.includes("/dist/")).shift() || "", c = s.startsWith("file://") ? x.fileURLToPath(s) : s, l = c.split(":")[0], u = await _.access(l).then(() => !0, () => !1);
		if (s && !this.sources.has(s) && u) {
			let e = await _.readFile(l, "utf-8");
			this.sources.set(l, e.toString()), this.sendUpstream("sources", { [l]: e.toString() });
		}
		let d = {
			command: t,
			args: n,
			result: r,
			error: i,
			timestamp: Date.now(),
			callSource: a ?? c
		};
		try {
			d.screenshot = await e.takeScreenshot();
		} catch (e) {
			k.warn(`failed to capture screenshot: ${e.message}`);
		}
		this.commandsLog.push(d), this.sendUpstream("commands", [d]), o.includes(t) && await this.#l(e);
	}
	async injectScript(e) {
		if (this.#t) {
			k.info("Script already injected, skipping");
			return;
		}
		if (!e.isBidi) throw new b(`Can not set up devtools for session with id "${e.sessionId}" because it doesn't support WebDriver Bidi`);
		this.#t = !0, k.info("Injecting devtools script...");
		let t = await se("@wdio/devtools-script", import.meta.url), n = `async () => { ${(await _.readFile(x.fileURLToPath(t))).toString()} }`;
		await e.scriptAddPreloadScript({ functionDeclaration: n }), k.info("✓ Script injected successfully");
	}
	async #l(e) {
		if (!this.#t) {
			k.warn("Script not injected, skipping trace capture");
			return;
		}
		try {
			if (!await e.execute(() => window.wdioTraceCollector !== void 0)) {
				k.warn("wdioTraceCollector not loaded yet - page loaded before preload script took effect");
				return;
			}
			let { mutations: t, traceLogs: n, consoleLogs: r, networkRequests: i, metadata: a } = await e.execute(() => window.wdioTraceCollector.getTraceData());
			if (this.metadata = a, Array.isArray(t) && (this.mutations.push(...t), this.sendUpstream("mutations", t)), Array.isArray(n) && (this.traceLogs.push(...n), this.sendUpstream("logs", n)), Array.isArray(r)) {
				let e = r;
				e.forEach((e) => e.source = h.BROWSER), this.consoleLogs.push(...e), this.sendUpstream("consoleLogs", e);
			}
			if (Array.isArray(i)) {
				let e = i;
				this.networkRequests.push(...e), this.sendUpstream("networkRequests", e);
			}
			this.sendUpstream("metadata", a), k.info(`✓ Sent metadata upstream, WS state: ${this.#e?.readyState}`);
		} catch (e) {
			k.error(`Failed to capture trace: ${e.message}`);
		}
	}
	handleNetworkRequestStarted(e) {
		try {
			let { request: t, timestamp: n } = e, r = t.request, i = {};
			t.headers && t.headers.forEach((e) => {
				let t = typeof e.name == "string" ? e.name.toLowerCase() : "", n = typeof e.value == "string" ? e.value : typeof e.value == "object" && e.value?.value ? e.value.value : "";
				t && (i[t] = n);
			}), this.#a.set(r, {
				url: t.url,
				method: t.method,
				timestamp: n,
				startTime: performance.now(),
				requestHeaders: i
			});
		} catch (e) {
			k.error(`handleNetworkRequestStarted error: ${e}`);
		}
	}
	handleNetworkResponseCompleted(e) {
		try {
			let { request: t, response: n, timestamp: r } = e, i = t.request, a = this.#a.get(i);
			if (!a) return;
			this.#a.delete(i);
			let o = {};
			n.headers && n.headers.forEach((e) => {
				let t = typeof e.name == "string" ? e.name.toLowerCase() : "", n = typeof e.value == "string" ? e.value : typeof e.value == "object" && e.value?.value ? e.value.value : "";
				t && (o[t] = n);
			});
			let s = o["content-type"]?.trim();
			if (!s || s === "-") return;
			let c = performance.now(), l = {
				id: `${r}-${i}`,
				url: a.url,
				method: a.method,
				status: n.status,
				statusText: n.statusText,
				type: this.#u(a.url, s),
				timestamp: a.timestamp,
				startTime: a.startTime,
				endTime: c,
				time: c - a.startTime,
				requestHeaders: a.requestHeaders,
				responseHeaders: o,
				size: n.bytesReceived
			};
			this.networkRequests.push(l), this.sendUpstream("networkRequests", [l]);
		} catch (e) {
			k.error(`handleNetworkResponseCompleted error: ${e}`);
		}
	}
	handleNetworkFetchError(e) {
		let t = e.request.request;
		this.#a.delete(t);
	}
	#u(e, t) {
		let n = e.toLowerCase(), r = t?.toLowerCase() || "";
		return r.includes("text/html") ? "document" : r.includes("text/css") ? "stylesheet" : r.includes("javascript") || r.includes("ecmascript") ? "script" : r.includes("image/") ? "image" : r.includes("font/") || r.includes("woff") ? "font" : r.includes("application/json") ? "fetch" : n.endsWith(".html") || n.endsWith(".htm") ? "document" : n.endsWith(".css") ? "stylesheet" : n.endsWith(".js") || n.endsWith(".mjs") ? "script" : n.match(/\.(png|jpg|jpeg|gif|svg|webp|ico)$/) ? "image" : n.match(/\.(woff|woff2|ttf|eot|otf)$/) ? "font" : "xhr";
	}
	sendUpstream(e, t) {
		!this.#e || this.#e.readyState !== S.OPEN || this.#e.send(JSON.stringify({
			scope: e,
			data: t
		}));
	}
}, N = T(import.meta.url), P = N("@babel/traverse").default, F = /* @__PURE__ */ new Map(), I;
try {
	let e = N("@cucumber/cucumber-expressions");
	I = {
		CucumberExpression: e.CucumberExpression,
		ParameterTypeRegistry: e.ParameterTypeRegistry
	};
} catch {}
var L;
function R(e) {
	L = e;
}
function z(e) {
	let t = e;
	return t.parent ? z(t.parent) : e;
}
function B(e) {
	if (e) {
		if (e.type === "Identifier") return e.name;
		if (e.type === "MemberExpression") {
			let t = e.object;
			return t && t.type === "Identifier" ? t.name : void 0;
		}
	}
}
function V(n) {
	if (!w.existsSync(n)) return [];
	let r = E(w.readFileSync(n, "utf-8"), {
		sourceType: "module",
		plugins: s,
		errorRecovery: !0,
		allowReturnOutsideFunction: !0
	}), i = [], a = [], o = (e) => !!e && t.includes(e) || e === "Feature", c = (t) => !!t && e.includes(t), l = (e) => {
		if (e) {
			if (e.type === "StringLiteral") return e.value;
			if (e.type === "TemplateLiteral" && e.expressions.length === 0) return e.quasis.map((e) => e.value.cooked).join("");
		}
	};
	return P(r, {
		enter(e) {
			if (!e.isCallExpression()) return;
			let t = e.node.callee, n = B(t);
			if (n) {
				if (o(n)) {
					let t = l(e.node.arguments?.[0]);
					t && (i.push({
						type: "suite",
						name: t,
						titlePath: [...a, t],
						line: e.node.loc?.start.line,
						column: e.node.loc?.start.column
					}), a.push(t));
				} else if (c(n)) {
					let t = l(e.node.arguments?.[0]);
					t && i.push({
						type: "test",
						name: t,
						titlePath: [...a, t],
						line: e.node.loc?.start.line,
						column: e.node.loc?.start.column
					});
				}
			}
		},
		exit(e) {
			if (!e.isCallExpression()) return;
			let t = e.node.callee, n = B(t);
			if (!n || !o(n)) return;
			let r = (() => {
				let t = e.node.arguments?.[0];
				if (t?.type === "StringLiteral") return t.value;
				if (t?.type === "TemplateLiteral" && t.expressions.length === 0) return t.quasis.map((e) => e.value.cooked).join("");
			})();
			r && a[a.length - 1] === r && a.pop();
		}
	}), i;
}
function H() {
	let e = C(/* @__PURE__ */ Error()), t = (t) => {
		let n = e.find((e) => {
			let n = e.getFileName();
			return !!n && !n.includes("node_modules") && t(e);
		});
		return n ? {
			file: n.getFileName(),
			line: n.getLineNumber(),
			column: n.getColumnNumber()
		} : null;
	};
	return t((e) => {
		let t = e.getFileName();
		return i.test(t) || oe.test(t);
	}) || t((e) => l.test(e.getFileName())) || t((e) => p.test(e.getFileName())) || null;
}
function U(e) {
	let t = e;
	for (let e = 0; e < 6; e++) {
		for (let e of c) {
			let n = v.join(t, e);
			if (w.existsSync(n) && w.statSync(n).isDirectory()) return n;
		}
		let e = v.dirname(t);
		if (e === t) break;
		t = e;
	}
}
var W;
function ue() {
	if (W && w.existsSync(W)) return W;
	let e = [{
		dir: process.cwd(),
		depth: 0
	}];
	for (; e.length;) {
		let { dir: t, depth: n } = e.shift();
		if (n > 5) continue;
		let r = v.join(t, "features");
		if (w.existsSync(r) && w.statSync(r).isDirectory()) for (let e of c) {
			let t = v.join(r, e);
			if (w.existsSync(t) && w.statSync(t).isDirectory()) return W = t, t;
		}
		for (let r of w.readdirSync(t)) {
			if (r.startsWith(".")) continue;
			let i = v.join(t, r), a;
			try {
				a = w.statSync(i);
			} catch {
				continue;
			}
			a.isDirectory() && !i.includes("node_modules") && e.push({
				dir: i,
				depth: n + 1
			});
		}
	}
}
function G(e) {
	let t = [];
	for (let n of w.readdirSync(e)) {
		let r = v.join(e, n);
		w.statSync(r).isDirectory() ? t.push(...G(r)) : te.test(n) && t.push(r);
	}
	return t;
}
function de(e) {
	let t = [], r = w.readFileSync(e, "utf-8").split(/\r?\n/);
	for (let i = 0; i < r.length; i++) {
		let a = r[i], o = a.match(n);
		if (o) {
			let n = o[2], r = n.lastIndexOf("/"), a = n.slice(1, r), s = n.slice(r + 1);
			try {
				t.push({
					kind: "regex",
					regex: new RegExp(a, s),
					file: e,
					line: i + 1,
					column: o.index ?? 0
				});
				continue;
			} catch {}
		}
		let s = a.match(ie);
		if (s) {
			let n = s[1], r = s[3];
			t.push({
				kind: "string",
				keyword: n,
				text: r,
				file: e,
				line: i + 1,
				column: s.index ?? 0
			});
		}
	}
	return t;
}
var K = /* @__PURE__ */ new Map();
function fe(e) {
	let t = K.get(e);
	if (t) return t;
	let n = G(e), r = [];
	for (let e of n) {
		let t = 0;
		try {
			P(E(w.readFileSync(e, "utf-8"), {
				sourceType: "module",
				plugins: s,
				errorRecovery: !0
			}), { CallExpression(n) {
				let i = n.node.callee, a;
				if (i?.type === "Identifier") a = i.name;
				else if (i?.type === "MemberExpression") {
					let e = i.property;
					e?.type === "Identifier" && (a = e.name);
				}
				if (!a || !ae.includes(a)) return;
				let o = n.node.arguments?.[0], s = {
					file: e,
					line: n.node.loc?.start.line ?? 1,
					column: n.node.loc?.start.column ?? 0
				};
				if (o?.type === "RegExpLiteral") r.push({
					kind: "regex",
					regex: new RegExp(o.pattern, o.flags ?? ""),
					...s
				}), t++;
				else if (o?.type === "StringLiteral") {
					if (I && o.value.includes("{")) {
						let e = new I.CucumberExpression(o.value, new I.ParameterTypeRegistry());
						r.push({
							kind: "expression",
							expr: e,
							...s
						});
					} else r.push({
						kind: "string",
						keyword: a,
						text: o.value,
						...s
					});
					t++;
				}
			} });
		} catch {}
		if (t === 0) {
			let t = de(e);
			t.length && r.push(...t);
		}
	}
	return K.set(e, r), r;
}
function pe(e, t) {
	let n = t ? v.extname(t) ? v.dirname(t) : t : void 0, r = n ? U(n) : void 0;
	if (r ||= ue(), !r) return;
	let i = fe(r), a = String(e ?? "").trim(), o = a.replace(/^(Given|When|Then|And|But)\s+/i, "").trim(), s = i.find((e) => e.kind === "string" && (o.localeCompare(e.text, "en", { sensitivity: "base" }) === 0 || a.localeCompare(`${e.keyword} ${e.text}`, "en", { sensitivity: "base" }) === 0));
	if (s) return {
		file: s.file,
		line: s.line,
		column: s.column
	};
	let c = i.find((e) => e.kind === "expression" && (() => {
		try {
			return !!e.expr.match(o) || !!e.expr.match(a);
		} catch {
			return !1;
		}
	})());
	if (c) return {
		file: c.file,
		line: c.line,
		column: c.column
	};
	let l = i.find((e) => e.kind === "regex" && (e.regex.test(o) || e.regex.test(a)));
	if (l) return {
		file: l.file,
		line: l.line,
		column: l.column
	};
}
function me(e) {
	return String(e || "").replace(/^\d+:\s*/, "").replace(/\s+/g, " ").trim();
}
function q(e) {
	return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function J(e, t) {
	let n = 1, r = 1;
	for (let i = 0; i < t && i < e.length; i++) e.charCodeAt(i) === 10 ? (n++, r = 1) : r++;
	return {
		line: n,
		column: r
	};
}
function he(t, n) {
	try {
		let r = w.readFileSync(t, "utf-8"), i = `(['"\`])${q(n)}\\1`, a = String.raw`\b(?:${e.join("|")})\s*\(\s*${i}`, o = new RegExp(a).exec(r);
		if (o && typeof o.index == "number") {
			let { line: e, column: n } = J(r, o.index);
			return {
				file: t,
				line: e,
				column: n
			};
		}
	} catch {}
}
function ge(e, n) {
	try {
		let r = w.readFileSync(e, "utf-8"), i = `(['"\`])${q(n)}\\1`, a = String.raw`\b(?:${t.join("|")})\s*\(\s*${i}`, o = new RegExp(a).exec(r);
		if (o && typeof o.index == "number") {
			let { line: t, column: n } = J(r, o.index);
			return {
				file: e,
				line: t,
				column: n
			};
		}
	} catch {}
}
function _e(e, t) {
	let n = String(e?.title ?? "").trim(), r = me(e?.fullTitle), i = (Array.isArray(e.specs) ? e.specs[0] : void 0) || e.file || e.specFile || t || L;
	if (/^(Given|When|Then|And|But)\b/i.test(n)) {
		let t = pe(n, p.test(String(i)) ? i : void 0);
		if (t) {
			Object.assign(e, t);
			return;
		}
	}
	let a = e.file || (Array.isArray(e.specs) ? e.specs[0] : void 0) || e.specFile || t || L;
	if (a && !p.test(a)) {
		if (!F.has(a)) try {
			F.set(a, V(a));
		} catch {}
		let t = F.get(a);
		if (t?.length) {
			let i = t.find((e) => e.type === "test" && e.name === n && r.includes(e.titlePath.join(" "))) || t.find((e) => e.type === "test" && e.name === n);
			if (i) {
				Object.assign(e, {
					file: a,
					line: i.line,
					column: i.column
				});
				return;
			}
		}
		let i = he(a, n);
		if (i) {
			Object.assign(e, i);
			return;
		}
	}
	let o = H();
	o && Object.assign(e, o);
}
function ve(e, t, n = []) {
	let r = String(e?.title ?? "").trim(), i = e.file || t || L;
	if (!r || !i) return;
	if (p.test(i)) {
		try {
			let t = w.readFileSync(i, "utf-8").split(/\r?\n/), n = (e) => e.trim().replace(/\s+/g, " "), a = n(r);
			for (let r = 0; r < t.length; r++) {
				let o = t[r].match(ne);
				if (o && n(o[2]) === a) {
					Object.assign(e, {
						file: i,
						line: r + 1,
						column: 1
					});
					return;
				}
			}
		} catch {}
		return;
	}
	try {
		F.has(i) || F.set(i, V(i));
		let t = F.get(i);
		if (t?.length) {
			let a = t.find((e) => e.type === "suite" && Array.isArray(e.titlePath) && e.titlePath.length === n.length && e.titlePath.every((e, t) => e === n[t])) || t.find((e) => e.type === "suite" && e.titlePath.at(-1) === r);
			if (a?.line) {
				Object.assign(e, {
					file: i,
					line: a.line,
					column: a.column
				});
				return;
			}
		}
	} catch {}
	let a = ge(i, r);
	a && Object.assign(e, a);
}
//#endregion
//#region src/reporter.ts
var Y = /* @__PURE__ */ new Map();
function X(e) {
	let t = e;
	if (t.type === "scenario" && /^\d+$/.test(t.uid)) {
		let n = [
			e.title,
			t.file || "",
			t.parent || "",
			t.cid || "",
			`example-${t.uid}`
		].join("::").split("").reduce((e, t) => (e << 5) - e + t.charCodeAt(0) | 0, 0);
		return `stable-${Math.abs(n).toString(36)}`;
	}
	let n = [t.file || "", String(t.fullTitle || e.title)], r = n.join("::"), i = Y.get(r) || 0;
	Y.set(r, i + 1), i > 0 && n.push(String(i));
	let a = n.join("::").split("").reduce((e, t) => (e << 5) - e + t.charCodeAt(0) | 0, 0);
	return `stable-${Math.abs(a).toString(36)}`;
}
function Z() {
	Y.clear();
}
function ye(e, t) {
	try {
		if (!D(e)) return null;
		let n = O(e, "utf-8").split("\n"), r = -1;
		for (let e = 0; e < n.length; e++) {
			let i = n[e].trim();
			if ((i.startsWith("Scenario Outline:") || i.startsWith("Scenario:")) && i.includes(t)) {
				r = e;
				break;
			}
		}
		if (r === -1) return null;
		let i = -1;
		for (let e = r; e < n.length; e++) if (n[e].trim().startsWith("Examples:")) {
			i = e;
			break;
		}
		if (i === -1) return null;
		let a = /* @__PURE__ */ new Map(), o = 0, s = !1;
		for (let e = i + 1; e < n.length; e++) {
			let t = n[e].trim();
			if (t.startsWith("Scenario") || t.startsWith("Feature:") || !t && o > 0) break;
			t.startsWith("|") && (s ? (a.set(o, e + 1), o++) : s = !0);
		}
		return a.size > 0 ? a : null;
	} catch (e) {
		return console.error("[Reporter] Failed to parse feature file:", e), null;
	}
}
var be = class extends ce {
	#e;
	#t;
	#n = [];
	constructor(e, t) {
		super(e), this.#e = t, Z();
	}
	onSuiteStart(e) {
		super.onSuiteStart(e);
		let t = e;
		if (t.type === "scenario" && e.file?.endsWith(".feature")) {
			let n = parseInt(t.uid, 10);
			if (!isNaN(n)) {
				let r = ye(e.file, e.title);
				if (r?.has(n)) {
					let i = r.get(n);
					t.featureFile = e.file, t.featureLine = i;
				}
			}
		}
		e.uid = X(e), this.#t = e.file, R(e.file), e.title && this.#n.push(e.title), ve(e, this.#t, this.#n), e.file && e.line !== null && (e.callSource = `${e.file}:${e.line}`), this.#r();
	}
	onTestStart(e) {
		super.onTestStart(e);
		let t = e;
		t.argument?.uri && typeof t.argument?.line == "number" && (t.featureFile = t.argument.uri, t.featureLine = t.argument.line), _e(e, this.#t), e.file && e.line !== null && (e.callSource = `${e.file}:${e.line}`), e.uid = X(e), this.#r();
	}
	onTestEnd(e) {
		super.onTestEnd(e), this.#r();
	}
	onSuiteEnd(e) {
		super.onSuiteEnd(e), e.title && this.#n[this.#n.length - 1] === e.title && this.#n.pop(), this.#n.length === 0 && (this.#t = void 0, R(void 0)), this.#r();
	}
	#r() {
		if (!this.suites) return;
		let e = [];
		for (let t of Object.values(this.suites)) if (t) {
			let n = t.uid;
			e.push({ [n]: t });
		}
		e.length > 0 && this.#e(e);
	}
	get report() {
		return this.suites;
	}
}, xe = re, Q = y("@wdio/devtools-service");
function Se(e) {
	let t = !1, n = new $();
	return n.captureType = g.Standalone, n.beforeSession(e, e), e.beforeCommand = Array.isArray(e.beforeCommand) ? e.beforeCommand : e.beforeCommand ? [e.beforeCommand] : [], e.beforeCommand.push(async function(e) {
		t || (t = !0, n.before(this.capabilities, [], this)), e === "deleteSession" && await n.after();
	}, n.beforeCommand.bind(n)), e.afterCommand = Array.isArray(e.afterCommand) ? e.afterCommand : e.afterCommand ? [e.afterCommand] : [], e.afterCommand.push(n.afterCommand.bind(n)), e;
}
var $ = class {
	constructor() {
		this.captureType = g.Testrunner;
	}
	#e = [];
	#t = new M();
	#n;
	#r = !1;
	#i = [];
	#a = null;
	#o = !1;
	async before(e, t, n) {
		this.#n = n, this.#t = new M(e["wdio:devtoolsOptions"]);
		try {
			await this.#s(n);
		} catch (e) {
			Q.error(`Failed to inject script at session start: ${e.message}`);
		}
		n.execute(() => window.visualViewport).then((e) => this.#t.sendUpstream("metadata", {
			viewport: e || void 0,
			type: this.captureType,
			options: n.options,
			capabilities: n.capabilities
		}));
	}
	beforeSession(e, t) {
		if (!("browserName" in t) && !("platformName" in t)) throw new b("The DevTools hook does not support multiremote yet");
		if ("reporters" in e) {
			let t = this;
			e.reporters = [...e.reporters || [], class extends be {
				constructor(e) {
					super(e, (e) => t.#t.sendUpstream("suites", e)), t.#e.push(this);
				}
			}];
		}
	}
	beforeScenario() {
		this.resetStack();
	}
	beforeTest() {
		this.resetStack();
	}
	resetStack() {
		this.#a = null, this.#i = [];
	}
	async beforeCommand(e, t) {
		if (!this.#n) return;
		!this.#r && this.#n.isBidi && (this.#r = !0, Q.info("Setting up BiDi network event listeners..."), this.#n.on("network.beforeRequestSent", (e) => {
			this.#t.handleNetworkRequestStarted(e);
		}), this.#n.on("network.responseCompleted", (e) => {
			this.#t.handleNetworkResponseCompleted(e);
		}), this.#n.on("network.fetchError", (e) => {
			Q.info(`>>> BiDi fetchError - keys: ${Object.keys(e).join(", ")}`), this.#t.handleNetworkFetchError(e);
		}), Q.info("✓ BiDi network event listeners registered")), e === "url" && this.#t.sendUpstream("metadata", { url: t[0] }), Error.stackTraceLimit = 20;
		let n = C(/* @__PURE__ */ Error("")).reverse().find((e) => {
			let t = e.getFileName();
			return t && ee.test(t);
		});
		if (n && this.#i.length === 0 && !a.includes(e)) {
			let r = n.getFileName() ?? void 0, i = r;
			if (r?.startsWith("file://")) try {
				let e = new URL(r);
				i = decodeURIComponent(e.pathname);
			} catch {
				i = r;
			}
			i?.includes("?") && (i = i.split("?")[0]);
			let a = n.getLineNumber() ?? void 0, o = n.getColumnNumber() ?? void 0, s = i === void 0 ? void 0 : `${i}:${a ?? 0}:${o ?? 0}`, c = JSON.stringify({
				command: e,
				args: t,
				src: s
			});
			this.#a !== c && (this.#i.push({
				command: e,
				callSource: s
			}), this.#a = c);
		}
	}
	afterCommand(e, t, n, r) {
		if (this.#o) return;
		let i = this.#i[this.#i.length - 1];
		if (i?.command === e && (this.#i.pop(), this.#n)) return this.#t.afterCommand(this.#n, e, t, n, r, i.callSource);
		u.includes(e) && this.#c(`context-change:${e}`);
	}
	async after() {
		if (!this.#n) return;
		let e = this.#n.options.outputDir || process.cwd(), { ...t } = this.#n.options, n = {
			mutations: this.#t.mutations,
			logs: this.#t.traceLogs,
			consoleLogs: this.#t.consoleLogs,
			networkRequests: this.#t.networkRequests,
			metadata: {
				type: this.captureType,
				...this.#t.metadata,
				options: t,
				capabilities: this.#n.capabilities
			},
			commands: this.#t.commandsLog,
			sources: Object.fromEntries(this.#t.sources),
			suites: this.#e.map((e) => e.report)
		}, r = v.join(e, `wdio-trace-${this.#n.sessionId}.json`);
		await _.writeFile(r, JSON.stringify(n)), Q.info(`DevTools trace saved to ${r}`), this.#t.cleanup();
	}
	async #s(e) {
		if (!e.isBidi) throw new b(`Can not set up devtools for session with id "${e.sessionId}" because it doesn't support WebDriver Bidi`);
		await this.#t.injectScript(z(e)), Q.info("✓ Devtools preload script active");
	}
	async #c(e) {
		if (!(!this.#n || this.#o)) try {
			if (this.#o = !0, await this.#n.execute(() => !!window.__WDIO_DEVTOOLS_MARK)) return;
			await this.#t.injectScript(z(this.#n));
		} catch (t) {
			Q.warn(`[inject] failed (reason=${e}): ${t.message}`);
		} finally {
			this.#o = !1;
		}
	}
};
//#endregion
export { g as TraceType, $ as default, xe as launcher, Se as setupForDevtools };
