// @wdio/devtools-service/dist/launcher-DpNc4Ovb.js
import e from "@wdio/logger";
import { remote as t } from "webdriverio";
import { start as n } from "@wdio/devtools-backend";
//#region src/constants.ts
var r = [
	"url",
	"navigateTo",
	"elementClick",
	"click"
], i = /\x1b\[[0-9;]*m/g, a = [
	"log",
	"info",
	"warn",
	"error"
], o = [
	{
		level: "trace",
		pattern: /\btrace\b/i
	},
	{
		level: "debug",
		pattern: /\bdebug\b/i
	},
	{
		level: "info",
		pattern: /\binfo\b/i
	},
	{
		level: "warn",
		pattern: /\bwarn(ing)?\b/i
	},
	{
		level: "error",
		pattern: /\berror\b/i
	}
], s = [
	"✗",
	"✓",
	"failed",
	"failure"
], c = {
	BROWSER: "browser",
	TEST: "test",
	TERMINAL: "terminal"
}, l = {
	browserName: "chrome",
	"goog:chromeOptions": { args: ["--window-size=1600,1200"] }
}, u = [
	"emit",
	"browsingContextLocateNodes",
	"browsingContextNavigate",
	"waitUntil",
	"getTitle",
	"getUrl",
	"getWindowSize",
	"setWindowSize",
	"deleteSession",
	"findElementFromShadowRoot",
	"findElementsFromShadowRoot",
	"waitForExist",
	"browsingContextGetTree",
	"scriptCallFunction",
	"getElement",
	"execute",
	"findElement",
	"getElementText",
	"getElementShadowRoot"
], d = [
	"url",
	"back",
	"forward",
	"refresh",
	"switchFrame",
	"newWindow",
	"createWindow",
	"closeWindow"
], f = /\/(test|spec|features|pageobjects|@wdio\/expect-webdriverio)\//i, p = [
	"typescript",
	"jsx",
	"decorators-legacy",
	"classProperties",
	"dynamicImport"
], m = [
	"it",
	"test",
	"specify",
	"fit",
	"xit"
], h = [
	"describe",
	"context",
	"suite"
], g = [
	"Given",
	"When",
	"Then",
	"And",
	"But",
	"defineStep"
], _ = /\.(?:steps?)\.[cm]?[jt]sx?$/i, v = /(?:^|\/)(?:step[-_]?definitions|steps)\/.+\.[cm]?[jt]sx?$/i, y = /\.(?:test|spec)\.[cm]?[jt]sx?$/i, b = /\.feature$/i, x = /\.(?:[cm]?js|[cm]?ts)x?$/, S = /^\s*(Feature|Scenario(?: Outline)?):\s*(.+)\s*$/i, C = /\b(Given|When|Then|And|But)\s*\(\s*(\/(?:\\.|[^/\\])+\/[gimsuy]*)/, w = /\b(Given|When|Then|And|But)\s*\(\s*(['`])([^'`\\]*(?:\\.[^'`\\]*)*)\2/, T = [
	"step-definitions",
	"step_definitions",
	"steps"
], E = e("@wdio/devtools-service:Launcher"), D = class {
	#e;
	#t;
	constructor(e) {
		this.#e = e;
	}
	async onPrepare(e, r) {
		try {
			let e = process.env.DEVTOOLS_APP_PORT, i = process.env.DEVTOOLS_APP_HOST || this.#e.hostname || "localhost";
			if (process.env.DEVTOOLS_APP_REUSE === "1" && e) {
				E.info(`Reusing existing DevTools app at http://${i}:${e}`), this.#n(r, {
					port: Number(e),
					hostname: i
				});
				return;
			}
			let { port: a } = await n({
				port: this.#e.port,
				hostname: this.#e.hostname
			});
			if (!a) return console.log(`Failed to start server on port ${a}`);
			this.#n(r, {
				port: a,
				hostname: this.#e.hostname || "localhost"
			}), this.#t = await t({
				automationProtocol: "devtools",
				capabilities: {
					...l,
					...this.#e.devtoolsCapabilities
				}
			}), await this.#t.url(`http://localhost:${a}`);
		} catch (e) {
			console.error(e);
		}
	}
	async onComplete() {
		if (this.#t) {
			for (e.setLevel("devtools", "warn"), E.info("Please close the browser window to finish...");;) try {
				await this.#t.getTitle(), await new Promise((e) => setTimeout(e, 1e3));
			} catch {
				E.info("Browser window closed, stopping DevTools app");
				break;
			}
			try {
				await this.#t.deleteSession();
			} catch (e) {
				E.warn("Session already closed or could not be deleted:", e.message);
			}
		}
	}
	#n(e, t) {
		if (Array.isArray(e)) for (let n of e) n["wdio:devtoolsOptions"] = {
			port: t.port,
			hostname: t.hostname || "localhost"
		};
	}
};
//#endregion
export { m as C, h as S, C as _, s as a, _ as b, u as c, r as d, p as f, T as g, y as h, d as i, o as l, f as m, i as n, b as o, x as p, a as r, S as s, D as t, c as u, w as v, g as x, v as y };
